### PR TITLE
Variable slab sizes

### PR DESCRIPTION
**Pull request by wthrowe**
## Proposed changes

There will be one more PR in this series that will add code to explicitly place slab boundaries so useful observations can be made without dense output.

Ping @geoffrey4444, because you expressed interest in this.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Time/Actions/ChangeSlabSize.hpp [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    auto& new_size_inbox =
+        tuples::get<ChangeSlabSize_detail::NewSizeInbox>(inboxes);
+    const double new_slab_size =
+        *alg::min_element(new_size_inbox.begin()->second);
+    new_size_inbox.erase(new_size_inbox.begin());
+
+    const TimeStepper& time_stepper =
+        Parallel::get<Tags::TimeStepperBase>(cache);
+
+    if (not time_stepper.can_change_step_size(
```
Comment (kidder) [inline src/Time/Actions/ChangeSlabSize.hpp:143]: Can you please add a comment why this check is done

File: src/Time/Actions/ChangeSlabSize.hpp [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+namespace ChangeSlabSize_detail {
+struct NewSizeInbox {
+  using temporal_id = int64_t;
+  using type = std::map<temporal_id, std::unordered_multiset<double>>;
+};
+
+struct ChangeTimesInbox {
+  using temporal_id = int64_t;
+  using NoData = std::tuple<>;
+  using type = std::map<temporal_id,
```
Comment (kidder) [inline src/Time/Actions/ChangeSlabSize.hpp:63]: comment on use of multiset (i.e. Event can be called multiple times>

Comment (kidder) [reply src/Time/Actions/ChangeSlabSize.hpp:63]: comment that this is effectively a counter...

File: src/Time/Actions/ChangeSlabSize.hpp [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+namespace Tags {
+struct DataBox;
+template <typename Tag>
+struct Next;
+}  // namespace Tags
+// IWYU pragma: no_forward_declare db::DataBox
+/// \endcond
+
+namespace ChangeSlabSize_detail {
+struct NewSizeInbox {
```
Comment (kidder) [inline src/Time/Actions/ChangeSlabSize.hpp]: NewSlabSizeInbox

File: src/Time/Actions/ChangeSlabSize.hpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+  using temporal_id = int64_t;
+  using type = std::map<temporal_id, std::unordered_multiset<double>>;
+};
+
+struct ChangeTimesInbox {
```
Comment (kidder) [inline src/Time/Actions/ChangeSlabSize.hpp]: NumberOfTimesCalledByEventInbox

File: src/Time/Actions/ChangeSlabSize.hpp [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+    auto& change_times_inbox =
+        tuples::get<ChangeSlabSize_detail::ChangeTimesInbox>(inboxes);
+    if (change_times_inbox.empty() or
+        change_times_inbox.begin()->first != time_step_id.slab_number()) {
+      return std::forward_as_tuple(std::move(box));
+    }
+    change_times_inbox.erase(change_times_inbox.begin());
+
+    auto& new_size_inbox =
```
Comment (kidder) [inline src/Time/Actions/ChangeSlabSize.hpp]: new_slab_size_inbox

File: src/Time/Actions/ChangeStepSize.hpp [hunk 6]
```diff
@@ -7,8 +7,6 @@
 #include <tuple>
```
Comment (kidder) [inline src/Time/Actions/ChangeStepSize.hpp:7]: typo in commit message `Those that care`

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp [hunk 7]
```diff
@@ -87,6 +89,23 @@ struct EvolutionMetavars {
   using limiter =
       Tags::Limiter<Limiters::Minmod<1, system::variables_tag::tags_list>>;
 
+  using step_choosers_common =
+      tmpl::list<StepChoosers::Registrars::Cfl<volume_dim, Frame::Inertial>,
+                 StepChoosers::Registrars::Constant,
+                 StepChoosers::Registrars::Increase>;
+  using step_choosers_step_only =
```
Comment (kidder) [inline src/Evolution/Executables/Burgers/EvolveBurgers.hpp]: ..._for_step_only . (and for_slab_only)

**Review by kidder — CHANGES_REQUESTED** [review]
joint review with @nilsdeppe and @wthrowe

**Comment by wthrowe** [comment]
I decided to go with NumberOfExpectedMessagesInbox instead of NumberOfTimesCalledByEventInbox.

**Comment by nilsdeppe** [comment]
review with @kidder LGTM, please squash

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**